c10/cuda/CUDAMallocAsyncAllocator.cpp

```diff
@@ -0,0 +1,887 @@
+}
+
+// This function should not issue any context-creating calls,
+// just set up for later calls to init per-device pools based
+// on the current device each later call sees.
+void init(int dev_count) {
+  static bool called = [](int dev_count) {
+    ;
+    // Are there external guarantees init will be called before
+    // any of the allocator's other functions?
+    // std::lock_guard<std::mutex> lk(general_mutex);
+    device_count = dev_count;
+    devs_initialized_flags.resize(dev_count, false);
+    dummy_unifying_free_streams.resize(dev_count);
+    pytorch_used_bytes.resize(dev_count);
+    pytorch_memory_limits.resize(dev_count);
+    return true;
+  }(dev_count);
+  (void)called;
+}
+
+static inline void assertValidDevice(int device) {
+  TORCH_CHECK(0 <= device && device < device_count, "Invalid device argument.");
+}
+
+void setMemoryFraction(double fraction, int device) {
+  TORCH_INTERNAL_ASSERT(
+      0 <= fraction && fraction <= 1,
+      "invalid fraction:",
+      fraction,
+      ". Please set within (0, 1).");
+
+  std::lock_guard<std::mutex> lk(general_mutex);
+  assertValidDevice(device);
+  CUDAGuard g(device);
+  // Should setMemoryFraction be allowed to trigger a full device context and
+  // pool-creating lazy_init_device, or should we simply assert this device is
+  // already initialized, ie
+  // TORCH_CHECK(devs_initialized_flags[device], ...)?
+  lazy_init_device(device);
+
+  size_t device_free;
+  size_t device_total;
+  C10_CUDA_CHECK(cudaMemGetInfo(&device_free, &device_total));
+  pytorch_memory_limits[device] =
+      static_cast<uint64_t>(fraction * device_total);
+
+  // Alternative: Instead of a manual hard limit, we could use
+  // cudaMemPoolSetAttribute(mempool, cudaMemPoolAttrReleaseThreshold,
+  // &threshold); This is a soft hint: The driver allows the pool's reserved
+  // memory to spike above threshold in regions of high cudaMallocAsync demand,
+  // but opportunistically trims reserved memory back to threshold when the
+  // memory in use is < threshold. I don't like this because it introduces
+  // performance nondeterminism.
+}
+
+void emptyCache(void) {
+  std::lock_guard<std::mutex> lk(general_mutex);
+
+  for (int dev = 0; dev < device_count; dev++) {
+    if (devs_initialized_flags[dev]) {
+      CUDAGuard g(dev);
+
+      cudaMemPool_t mempool;
+      cudaDeviceGetDefaultMemPool(&mempool, dev);
+      cudaDeviceSynchronize();
+      cudaMemPoolTrimTo(mempool, 0);
+    }
+  }
+}
+
+void cacheInfo(int device, size_t* maxWorkspaceGuess) {
+  // The only consumer of cacheInfo is getMaxWorkspaceSize in Conv_v7.cpp.
+  // Afaict, the role of cacheInfo is to give getMaxWorkspaceSize a reasonable
+  // maximum workspace size to use for an upcoming cudnnFind call.
+  //
+  // The native allocator's cacheInfo chooses to return the size of its largest
+  // unused block (which is the largest allocation the native allocator can
+  // service immediately and asynchronously without a cudaMalloc.
```

Comment:
nit: closing parenthesis after cudaMalloc, or just replace the opening parenthesis with a comma